include/exadg/incompressible_navier_stokes/spatial_discretization/spatial_operator_base.cpp

```diff
@@ -249,20 +249,87 @@ template<int dim, typename Number>
+    // degree - 1 is used so that the same Gauss quadrature can be used for exact integration.
+    fe_u = std::make_shared<dealii::FE_RaviartThomasNodal<dim>>(param.degree_u - 1);
+
+    pcout << std::endl
+          << "Hdiv-conforming Raviart-Thomas finite element discretization:" << std::endl
+          << "(Note that the given degree for the velocity is the degree in normal direction."
+          << std::endl
+          << " The degree in tangent direction is degree - 1. Note also that the number of dofs"
+          << std::endl
+          << "given by the dof_handler does not take the constraint of periodic boundaries into"
+          << std::endl
+          << " account.)" << std::endl
+          << std::endl
+          << std::flush;
+  }
+  else
+    AssertThrow(false, dealii::ExcMessage("FE not implemented."));
+
   // enumerate degrees of freedom
   dof_handler_u.distribute_dofs(*fe_u);
   dof_handler_p.distribute_dofs(fe_p);
   dof_handler_u_scalar.distribute_dofs(fe_u_scalar);
 
+  if(param.spatial_discretization == SpatialDiscretization::HDIV)
+  {
+    // We need to make sure the normal dofs are shared between cells on the periodic boundaries,
+    // since these are contiguous for HDIV.
+    dealii::IndexSet relevant_dofs;
+    dealii::DoFTools::extract_locally_relevant_dofs(dof_handler_u, relevant_dofs);
+    constraint_u.reinit(relevant_dofs);
+
+    auto const & periodic_face_pair = grid->periodic_faces;
+    for(auto const & face : periodic_face_pair)
+      dealii::DoFTools::make_periodicity_constraints(
+        dof_handler_u,
+        face.cell[0]->face(face.face_idx[0])->boundary_id(),
+        face.cell[1]->face(face.face_idx[1])->boundary_id(),
+        face.face_idx[0] / 2,
+        constraint_u);
+
+
+    // Set boundaries constraints
+    if(boundary_descriptor->velocity->symmetry_bc.empty() == false)
+    {
+      for(auto bc : boundary_descriptor->velocity->symmetry_bc)
+        dealii::VectorTools::project_boundary_values_div_conforming(
+          dof_handler_u, 0, *(bc.second), bc.first, constraint_u, *get_mapping());
+    }
+    if(boundary_descriptor->velocity->dirichlet_bc.empty() == false)
+    {
+      AssertThrow(
+        false,
+        dealii::ExcMessage(
+          "Dirichlet BC are not properly implemented. The BC needs to be strongly applied in order to obtain an exactly divergence-free solution."));
+      // We would like to do something similar to above. Probably with
+      // dealii::VectorTools::interpolate_boundary_values
+      // dealii::VectorTools::project_boundary_values
+      // dealii::DoFTools::make_zero_boundary_constraints
+    }
+  }
+
   unsigned int const ndofs_per_cell_velocity =
-    dealii::Utilities::pow(param.degree_u + 1, dim) * dim;
+    (get_spatial_discretization() == SpatialDiscretization::L2) ?
+      dealii::Utilities::pow(param.degree_u + 1, dim) * dim :
+      dealii::Utilities::pow(param.degree_u, dim - 1) * (param.degree_u + 1) * dim;
```

Comment:
I am not sure whether the ``... ? ... : ...`` syntax is suitable here. I think we could be more explicit using `if(L2)`, `else if (Hdiv)`, `else {AssertThrow(false, ExcNotImplemented);}`.

For the Hdiv-case, an additional comment like "continuity constraints in normal direction are not taken into account regarding the number of DoFs per element.".

Reply:
The second paragraph is optional, so I mark this comment as resolved.

Reply:
resolved